### DISCUSSION TITLE
Convert from HTML?

### DISCUSSION THREAD

**Discussion by philiprenich** [stepwisehq/prosemirror-dotnet#24] [Q&A] [answered]
Is it possible to convert from HTML? Prosemirror has DOMParser and while I realise that's different than a raw HTML string, it would be really useful. I suppose it's a harder problem outside of JS where there is a native DOM API. 

Alongside that, did I miss a way to take the schema structure and output an HTML string? That is a useful thing for a backend to do that's serving up documents to the browser.

Thanks!

**Comment by ProTip** [comment, marked as answer]
Hi @philiprenich ,

Currently there is no support for converting between HTML. There are a number of good C# libraries like [AngleSharp](https://github.com/AngleSharp/AngleSharp) that could be used to implement this.

However, it's complicated by the fact that many(most?) schemas use javascript as part of the HTML parsing/generation. This could be worked around with something like [Jint](https://github.com/sebastienros/jint) or [ClearScript](https://github.com/microsoft/ClearScript).

The lift is pretty big though, and I'm not sure I would get around to it unless it were sponsored work and/or I was hired to work on it 😆

**Reply by philiprenich** [reply]
Hi @ProTip , thanks for the reply and pointing me toward AngleSharp. My usecase only needs to go from HTML -> JSON once (porting over old content) so that should be feasible.

When you say schemas use JS as part of the HTML parsing, are you talking about the customisations developers create for how various elements get rendered?

When I was exploring this last week it did indeed look like a big effort, so good call on doing it through sponsorship! Sadly, I'm just a lone dev with a small need. Again, thanks for getting me looking in the right direction!

**Reply by ProTip** [reply]
`NodeSpec` reference: https://prosemirror.net/docs/ref/#model.NodeSpec .

The `NodeSpec` in the `schema` contains the fields `toDOM` and `parseDOM`. Each of these can end up containing javascript functions(`toDOM` directly and `parseDOM` in the parse rules).

Currently these are _unimplemented_ in `ProseMirror.Net` (the fields are actually left out IIRC). To fully support correctly converting between HTML for _any valid schema_ the JavaScript functions would likely need to be taken into account.

Even then, there would likely need to be constraints on the APIs those functions could call if the code were run outside the browser(via Jint or Clearscript). If one were to embed Chrome via CefSharp that could open up more options and less constraints.

**Reply by ProTip** [reply]
Depending on your needs, it may make sense to do the conversion via NodeJS or in the Browser.

**Reply by philiprenich** [reply]
I would think the contents for those `toDOM`/`parseDOM` implementations in a package like this would be left up to the developer, no? If they're using Prosemirror (or something built on top of it like TipTap) it would be up to them to correctly port their nodes' custom rendering/parsing functions to C#. 

(I *hope* I'm understanding the PM library correctly!)

Still, you're right that it may be very difficult/impossible to match what is done in the front-end due to API constraints from the browser.

**Reply by ProTip** [reply]
Adding HTML support to `ProseMirror.Net` would probably involve implementing [DOMParser](https://prosemirror.net/docs/ref/#model.DOMParser) and [DOMSerializer](https://prosemirror.net/docs/ref/#model.DOMSerializer).

If the developer were expected to re-implement `toDOM` and `parseDOM` javascript functions as C# functions that would certainly lessen the cost and complexity of implementation 🤔

For compatibility with existing schemas without lots of extra work, finding a way to execute the JavaScript code would be ideal. Even a lot of the basic Node types use functions to parse elements into nodes: https://github.com/ProseMirror/prosemirror-schema-basic/blob/master/src/schema-basic.ts#L86 .

**Reply by philiprenich** [reply]
Right, and those two classes aren't going to be simple to port. Something like AngleSharp may help (haven't delved into it too much yet). But with what you're saying with basic (almost default) nodes, you'd end up going in the direction that the official PHP TipTap library does where they've created all those Node types as their own extensions.

Well, for what I need at the moment, a NodeJS data port to get me into PM schema should be enough, and then this library for ongoing manipulation. If extra requirements come up down the line, I may investigate further - maybe see if I can extend parsing/serialising on top or see if you'd consider a PR to add it in here.

Thanks!

**Comment by philiprenich** [comment]
@ProTip So I've spent some time looking at this problem and have a working DOM Serializer (`ToDom`). Would you be interested in a PR to include it in this repo? I'll need to do some cleanup and more thoughtful code before submission, but wanted to check in before I do so.

Given the complexity, I was thinking doing this piecemeal and only adding the JSON->DOM direction for now and then look at what it would take to parse markup to turn it into PM schema JSON.

**Reply by ProTip** [reply]
Hi @philiprenich,

Sorry for the delay; work and travel got me(I can't even believe this post is nearly a month old 😞).

Yes, I'd love to take a look at a PR and I was actually getting interested in taking stab at this based on our previous discussion 😆. I might be overly opinionated, but at a minimum I can link to your project.

As an aside I'm hoping to open source the StepWise project this was created for soon.. That will have a concrete, full-stack implementation of a versioned document system built with this and `prosemirror-collab-commit`..

**Reply by philiprenich** [reply]
No worries, that's how it goes. I'm also about to be travelling - that will either give me no time to get the PR sorted or plenty of downtime for it! I think it's mostly tidy now. Caveat that C# isn't my primary language and I was trying to just replicate the Typescript code from Prosemirror (it looked like this library tried to stay close to it as well and be a direct port), but there are a few things that may not make sense. Similar stuff you alluded to before with the difference in browser APIs. All that to say, hopefully it's a good starting point (it does work!), but an opinionated look would be great.

Will see what I can do to get a PR up soon.

**Reply by philiprenich** [reply]
@ProTip Whew! Finally go there! Please see https://github.com/stepwisehq/prosemirror-dotnet/pull/25

Cheers!
